packages/adapters/watcher/src/verifiers/proposal.ts

```diff
@@ -0,0 +1,249 @@
+   *
+   * @returns object - Whether the invariant was verified. `true` if propose was verified (no switch needed).
+   * `false` if the invariant was violated (switch is needed)! If false, a reason is included in the response object.
+   */
+  public override async checkInvariant(requestContext: RequestContext): Promise<VerifyResponse> {
+    // get root manager contract of the hub domain chain
+    const hubDomainChainId = domainToChainId(this.hubDomain);
+    const rootManagerContract = this.getRootManagerContract(hubDomainChainId);
+
+    // get proposed data from root manager contract
+    const proposedData = await this.getProposedData(rootManagerContract, hubDomainChainId, requestContext);
+    const { aggregateRoot: proposedRoot } = proposedData;
+
+    // stop check execution if the proposed root dispute period has elapsed
+    const isDisputeTimeOver = await this.checkDisputeTimeFinalization(+proposedData.endOfDispute);
+    if (isDisputeTimeOver) {
+      return {
+        needsAction: false,
+        reason: `Dispute time for root ${proposedRoot} is already over`,
+      };
+    }
+
+    // get snapshot roots from spoke connectors and insert them into the virtual tree to generate the virtual root
+    const spokeConnectors = proposedData.domains.map((domain) => this.getSpokeConnectorData(+domain));
+    const snapshotRoots = await this.getSnapshotRoots(proposedData, spokeConnectors);
+    const virtualRoot = await this.insertVirtualLeaves(snapshotRoots);
+
+    // Invariant: proposedRoot == virtualRoot
+    // Compare virtual root with proposed root and return if action is needed or not
+    if (virtualRoot !== proposedRoot) {
+      return {
+        needsAction: true,
+        reason: `proposedRoot: ${proposedRoot} is different than virtualRoot: ${virtualRoot}`,
+      };
+    }
+    return {
+      needsAction: false,
+      reason: "Proposed root matches virtual root",
+    };
+  }
+
+  /**
+   * @notice Get the latest proposed aggregate root in the RootManager contract at a given snapshot.
+   * @param rootManager - Root Manager contract instance.
+   * @param hubDomainChainId - The hub domain chain id.
+   * @param requestContext - The request context.
+   * @returns ProposedData representing the data emitted by the last AggregateRootProposed event.
+   */
+  public async getProposedData(
+    rootManager: Contract,
+    hubDomainChainId: number,
+    requestContext: RequestContext,
+  ): Promise<ProposedData> {
+    const { methodContext } = createLoggingContext(this.getProposedData.name);
+
+    // It's important to ensure the provider doesn't have a block-limit to fetch block from, otherwise we must add a contraint
+    // in the toBlock and fromBlock optional parameters of rootManagerInstance.queryFilter
+
+    try {
+      // create a filter to get the latest proposed root event
+      const eventFilter = rootManager.filters.AggregateRootProposed();
+      const events = await rootManager.queryFilter(eventFilter);
```

Comment:
this will need some block boundaries otherwise some providers will error (and most all will be slow)